File: single-element-in-a-sorted-array/single-element-in-a-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNonDuplicate = function(nums) {
    var l = 0;
    var r = nums.length-1;
    while(l<r)
    {
        let m = l + Math.floor((r - l)/2);
        if(nums[m] == nums[m ^ 1])
            l = m +1;
        else
            r = m;
    }
    return nums[l];
};

File: single-element-in-a-sorted-array/single-element-in-a-sorted-array.cpp
class Solution {
public:
    //xor
    int singleNonDuplicate1(vector<int>& nums) {
        int ans(0);
        
        for(const auto& n:nums)
        {
            ans ^=n;
        }
        return ans;
    }
    // O(log n)
    int singleNonDuplicate(vector<int>& nums) {
        int l = 0;
        int r = nums.size()-1;
        
        while(l<r)
        {
            int mid = l + (r-l)/2;
            if(nums[mid] == nums[mid ^ 1])
                l = mid +1;
            else
                r = mid;
        }
        return nums[l];
    }
};